federation/pkg/federation-controller/service/endpoint_helper.go

```diff
@@ -31,22 +31,42 @@ import (
 // worker runs a worker thread that just dequeues items, processes them, and marks them done.
 // It enforces that the syncHandler is never invoked concurrently with the same key.
 func (sc *ServiceController) clusterEndpointWorker() {
-	fedClient := sc.federationClient
+	// process all pending events in endpointWorkerDoneChan
+	eventPending := true
+	for eventPending {
+		select {
+		case clusterName := <-sc.endpointWorkerDoneChan:
+			sc.endpointWorkerMap[clusterName] = false
+		default:
+			// non-blocking, comes here if all existing events are processed
+			eventPending = false
+			break
+		}
+	}
+
 	for clusterName, cache := range sc.clusterCache.clientMap {
+		workerExist, keyFound := sc.endpointWorkerMap[clusterName]
+		if keyFound && workerExist {
+			continue
+		}
+		sc.endpointWorkerMap[clusterName] = true
+
+		// create a worker only if the previous worker has finished and gone out of scope
 		go func(cache *clusterCache, clusterName string) {
+			fedClient := sc.federationClient
 			for {
-				func() {
```

Comment:
This function is important here.

Notice the defer statement inside this block. This `func()` exists for that sole reason. At the end of each loop, we need to say we are done for that key, so we `defer cache.endpointQueue.Done(key)`. If you remove the `func() {}` wrapping, defer will be run at the end of the enclosing goroutine and the keys won't be removed from the queue until then. So I don't think you should be removing the `func() {}` wrapping.